### PR TITLE
[Merged by Bors] - add more `SAFETY` comments and lint for missing ones in `bevy_ecs`

### PR DESCRIPTION
**Pull request by jakobhellermann**
# Objective

`SAFETY` comments are meant to be placed before `unsafe` blocks and should contain the reasoning of why in this case the usage of unsafe is okay. This is useful when reading the code because it makes it clear which assumptions are required for safety, and makes it easier to spot possible unsoundness holes. It also forces the code writer to think of something to write and maybe look at the safety contracts of any called unsafe methods again to double-check their correct usage.

There's a clippy lint called `undocumented_unsafe_blocks` which warns when using a block without such a comment. 

## Solution

- since clippy expects `SAFETY` instead of `SAFE`, rename those
- add `SAFETY` comments in more places
- for the last remaining 3 places, add an `#[allow()]` and `// TODO` since I wasn't comfortable enough with the code to justify their safety
- add ` #![warn(clippy::undocumented_unsafe_blocks)]` to `bevy_ecs`

### Note for reviewers

The first commit only renames `SAFETY` to `SAFE` so it doesn't need a thorough review.
https://github.com/bevyengine/bevy/pull/4835/files/cb042a416ecbe5e7d74797449969e064d8a5f13c..55cef2d6fa3aa634667a60f6d5abc16f43f16298 is the diff for all other changes.

### Safety comments where I'm not too familiar with the code

https://github.com/bevyengine/bevy/blob/774012ece50e4add4fcc8324ec48bbecf5546c3c/crates/bevy_ecs/src/entity/mod.rs#L540-L546

https://github.com/bevyengine/bevy/blob/774012ece50e4add4fcc8324ec48bbecf5546c3c/crates/bevy_ecs/src/world/entity_ref.rs#L249-L252

### Locations left undocumented with a `TODO` comment

https://github.com/bevyengine/bevy/blob/5dde944a3051426ac69fdedc5699f7da97a7e147/crates/bevy_ecs/src/schedule/executor_parallel.rs#L196-L199

https://github.com/bevyengine/bevy/blob/5dde944a3051426ac69fdedc5699f7da97a7e147/crates/bevy_ecs/src/world/entity_ref.rs#L287-L289

https://github.com/bevyengine/bevy/blob/5dde944a3051426ac69fdedc5699f7da97a7e147/crates/bevy_ecs/src/world/entity_ref.rs#L413-L415

### CODE REVIEW

File: crates/bevy_ecs/src/schedule/executor_parallel.rs
Comment (hymm) [inline crates/bevy_ecs/src/schedule/executor_parallel.rs:196]:
```suggestion
                    // SAFETY: the executor prevents two systems with conflicting access from running simultaneously.
```

Comment (hymm) [reply crates/bevy_ecs/src/schedule/executor_parallel.rs:196]: it's the call to this function https://github.com/bevyengine/bevy/blob/main/crates/bevy_ecs/src/schedule/executor_parallel.rs#L244

**Review by alice-i-cecile — COMMENTED** [review]
This should note that Command: Send as a trait bound.

**Review by alice-i-cecile — APPROVED** [review]
I'm happy with both the direction and the safety comments added. I think this should wait on @cart as it's a) a change to project style and b) mostly their code.

**Comment by cart** [comment]
Definitely on board for these changes! I'll done one last review once conflicts have been resolved, but I think this is pretty much good to go.

**Review by cart — APPROVED** [review]
The new safety docs are correct and checking for undocumented unsafe blocks is a huge improvement. Thanks!

One last round of merge conflict resolutions and then we should merge this asap.

**Review by alice-i-cecile — APPROVED** [review]
New changes LGTM. The runtime check is worth it for soundness too. I'll retry this once I get a second approval.

**Review by james7132 — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

**Pull request closed without merging**
